yellow-paper/docs/public-vm/avm-circuit.md

```diff
@@ -3,83 +3,137 @@ sidebar_position: 1
-    - flags: `callPointer`, `userMemory = 1` (user memory access)
-    - offset: `dstOffset`
+> Refer to ["AVM State Model"](./state-model) for more details on the absence of "external registers" in the AVM.
 
-## Circuit I/O
+## Control Flow Unit
+Processes updates to the program counter and call pointer to ensure that execution proceeds properly from one instruction to the next.
+
+### Program Counter
+The Control Flow Unit's **program counter** is an index into the bytecode of the current call's contract. For most instructions, the Control Flow Unit will simply increment the program counter. Certain instructions (like `JUMP`) decode into control flow sub-operations (like `PCSTORE`). The Control Flow Unit processes such instructions to update the program counter.
+
+### Call Pointer
+A **message call pointer** uniquely identifies a message call among all message calls processed by the current circuit. The Control Flow Unit tracks the currently active call pointer and the next available one. When a nested message call is encountered, it assigns it the next available call pointer (`callPointer = nextCallPointer++`) and increments that next pointer value. It then sets the program counter to 0.
+
+A request's initial message call is assigned call pointer of `1`. The first nested message call encountered during execution is assigned call pointer of `2`. The Control Flow Unit assigns call pointers based on execution order.
+
+There is certain information that must be tracked by the VM circuit on a per-call basis. For example, each call will correspond to the execution of a different contract's bytecode, and each call will access call-specific memory. As a per-call unique identifier, the message call pointer enables bytecode and memory lookups, among other things, on a per-call basis.
+
+#### "Input" and "output" call pointers
+It is important to note that the initial message call's pointer is `1`, not `0`. The zero call pointer is a special case known as the "input" call pointer.
+
+As expanded on later, the VM circuit memory table has a separate section for each call pointer. The memory table section for the **input call pointer** is reserved for the initial call's `calldata`. This will be expanded on later.
+
+### Internal Call Stack
+**TODO**
+### Nested Message Calls
+**TODO**
+#### Initializing nested call context
+**TODO**
+#### Snapshotting and restoring context
+**TODO**
+
+## Memory Controller
+The VM circuit's **Memory Controller** processes loads and stores between intermediate registers and memory.
 
+### Memory Sub-operations
+When decoded, instructions that operate on memory map to some Memory Controller sub-operations. A memory read maps to a `LOAD` sub-operation which loads a word from memory into an internal register. The memory offset for this sub-operation is generally specified by an instruction argument. Similarly, a memory write maps to a `STORE` sub-operation which stores a word from an internal register to memory.
```

Comment:
Use "intermediate register" terminology rather than "internal register". (two occurences in this paragraph)